core/templates/dev/head/base-components/base-content.directive.ts

```diff
@@ -40,10 +40,20 @@ angular.module('oppia').directive('baseContent', [
       templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
         '/base-components/base-content.directive.html'),
       controllerAs: '$ctrl',
-      controller: ['$rootScope', 'BackgroundMaskService',
+      controller: ['$rootScope', '$window', 'BackgroundMaskService',
         'SidebarStatusService', 'UrlService', 'SITE_FEEDBACK_FORM_URL',
-        function($rootScope, BackgroundMaskService,
+        function($rootScope, $window, BackgroundMaskService,
             SidebarStatusService, UrlService, SITE_FEEDBACK_FORM_URL) {
+          // Mimic redirection behaviour in the backend (see issue #7867 for
+          // details).
+          if ($window.location.hostname === 'oppiaserver.appspot.com') {
+            $window.location.href = (
+              'https://oppiatestserver.appspot.com' +
+              $window.location.pathname +
+              $window.location.search +
+              $window.location.hash);
```

Comment:
`$window.location.hostname = 'oppiatestserver.appspot.com'` should be enough here, it replaces just the hostname part and automatically keeps the others.

Reply:
I just tried doing that, but initially I found that the page doesn't redirect, and the URL does not change, although the if condition is triggered.

Subsequently I also found that this didn't work quite as expected because, despite the redirect, port names were preserved as well.

So, would prefer to leave it constructed explicitly as-is. Would that be OK?

Reply:
I think there shouldn't be any difference in reloading when setting `href` or `hostname`, but it could be some random coincidence.

Okay, if we don't want to preserve the port number we can do it this way, but I'm thinking about maybe moving this code to *services/contextual/url.service.ts* and also writing some FE tests for it.

Reply:
Actually, I don't see the code I wrote for the URL being reusable (and I don't want to encourage that TBH). How strongly do you feel about moving it to url.service?

Reply:
Okay, that makes sense.